### ISSUE TITLE
Replacing subscriptions for testing

### ISSUE THREAD

**Issue by hschne** [rmosolgo/graphql-ruby#2687]
We are facing an issue where we would like to replace a custom subscriptions object on the GraphQL Schema on the fly  (mostly for testing). We currently have the following schema: 

```ruby
class MySchema < GraphQL::Schema
  query CustomQueryType
  mutation CustomMutationType
  subscription CustomSuscriptionType

  use CustomSubscriptions # We want to replace this on the fly
end
```

`CustomSubscriptions` itself derives from `GraphQL::Subscriptions` and uses some external dependencies to store/trigger subscriptions. While testing, we don't necessarily want to deal with these dependencies, so we would like to replace the subscriptions there with some mock or some implementation that does not require the same external dependencies. For example, I'd like to do: 

```ruby
class SubscriptionsTest < ActiveSupport::TestCase
 def setup
   mySchema.use(MyTestSubscriptions) # This doesn't work :(
 end

 ...
end
```

However, that will not remove our `CustomSubscriptions` from the schema. I have tried to update the schema to allow setting the subscriptions instance, but so far with no success. 

Is there a way to accomplish what I'm after? Some setter/helper method that I missed?

**Comment by rmosolgo** [comment]
Hmm, thanks for the great writeup. 

> so far with no success.

What happened when you tried it? 

In GraphQL 1.9, I'd _expect_ that the problem has to do with some hidden caching inside the schema. For compatibility reasons, each class-based schema _also_ generates and caches a `GraphQL::Schema` instance. Updating the class _doesn't_ update the generated instance. I've been working hard to remove that layer of caching and I just released 1.10 which doesn't use caching for class-based schemas which use the new `Interpreter` runtime.

Using 1.10, you could subclass the main schema, and use test subscriptions for the subclass, eg 

```ruby 
test_schema = Class.new(MySchema) { use(TestSubscriptions) }
```

That _should_ work!

Using 1.9, you could use `.redefine` to return a cloned _instance_ of the cached schema, which should work exactly the same as the schema instance: 

```ruby 
test_schema = MySchema.redefine { use(TestSubscriptions }
```

But, the resulting `test_schema` will not be compatible with the `Interpreter` in 1.10, which requires class-based schemas. 

Want to try one of those approaches?

**Comment by hschne** [comment]
> Using 1.10, you could subclass the main schema, and use test subscriptions for the subclass

I tried that. The problem is that other parts of the code use our original schema, and I simply can't replace all the instances where that is used with the new test schema. 

Redefined did not work for me for the same reason. 

> Updating the class doesn't update the generated instance

That got me on the right track. Using the instance and modifying `subscriptions` directly there, I can make it work. The following code replaces the subscriptions instance globally and can also be used to reset it to the previous one: 

```
@subscriptions  = MyTestSubscription.new(schema: MeistertaskSchema.graphql_definition)
MeistertaskSchema.graphql_definition.subscriptions = @subscriptions
```

Thanks for the quick turnaround!
